superset-embedded-sdk/CONTRIBUTING.md

```diff
@@ -0,0 +1,57 @@
+<!--
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements.  See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+
+# Contributing to the Superset Embedded SDK
+
+The superset-embedded-sdk directory is a self contained sub-project in the Superset codebase.
+
+This is because the SDK has different requirements from other parts of the Superset codebase:
+Namely, we need to export a lightweight frontend library that can be used in as many environments as possible.
+Having separate configs allows for better separation of concerns and allows the SDK code to remain simple.
+
+## Building
+
+The library is built in two modes: one for consumption by package managers
+and subsequent build systems, and one for consumption directly by a web browser.
+
+Babel is used to build the sdk into a relatively modern js package in the `lib` directory.
+This is used by consumers who install the embedded sdk via npm, yarn, or other package manager.
+
+Webpack is used to bundle the `bundle` directory,
+for use directly in the browser with no build step e.g. when importing via unpkg.
+
+Typescript outputs type definition files to the `dist` directory.
+
+Which of these outputs is used by the library consumer is determined by our package.json's `main`, `module`, and `types` fields.
+
+## Testing
+
+You may notice a lack of tests in this directory. The functions used in the sdk so far are very closely tied to browser behavior,
```

Comment:
I'm not sure about this. It would be really cool if we could spend more time trying to use Jest/JSDOM to write tests that check for specific conditions like:
- Failure when invoking `fetchGuestToken`
- An unknown `id`
- An invalid `supersetDomain`
- Test the unmount behavior

You could use `"testEnvironmentOptions": { "resources": "usable" }` in `jest.config.js` to enable [sub-resouces](https://github.com/jsdom/jsdom#loading-subresources) like iframes.

Reply:
Hmm, I think you're right that it would be good to unit test the `fetchGuestToken` handling. And it would be nice to at least have one test to ensure that the code doesn't blow up. As for the others though, in my experience, unit testing helps most when you have branching logic in your code, and this code currently doesn't contain even one if statement. It just executes a series of instructions, all of which are interactions with external APIs. So, unit testing it beyond the "it doesn't blow up" case will essentially be asserting that the code exists as written. Those kinds of tests tend to just get in the way of changing the code instead of really helping much.

I'll also add an "on pull request" github action that makes sure the build can still run.

Reply:
For case 1, the SDK does not have any clue that the id doesn't exist. All it does is create an iframe and send a guest token to it.

In the second case, what happens is entirely dependent on browser implementation. Some browsers could throw an error, which would be fine imo since the caller passed in the wrong thing it should probably throw anyway, but some browsers silently ignore the message, I believe this is to prevent a malicious page from gathering info about other pages.

Reply:
In the case a browser silently ignores the message, won't this line fail later?
```
ourPort.postMessage({ guestToken });
```

Reply:
It will fail to actually convey the guest token to its intended recipient since no one is listening on the other side, but it won't throw an error. The message channel and port do exist, they're just not connected to anything.

Reply:
I wrote a test for "it doesn't blow up", but actually jsdom doesn't seem to support the iframe's `sandbox` property. It's just not present on the iframe element. Unless someone knows of a good way around that, I think I'll forego unit testing in favor of thorough cypress testing using the demo host app, which should be added soon. For now, this code has been manually tested pretty thoroughly in a couple of applications, and we're focused on getting it usable asap.